diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/BasicFunctions/FFT.glsl 0.8.22-patched/shaders/Lib/BasicFunctions/FFT.glsl
--- 0.8.22-orig/shaders/Lib/BasicFunctions/FFT.glsl	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/Lib/BasicFunctions/FFT.glsl	2026-05-26 19:07:18.131237500 -0700
@@ -0,0 +1,153 @@
+// by huj31415
+
+#define PI 3.1415926535
+#define TAU 6.283185307
+
+#if FFT_SIZE == 1024
+  #define N_4_1 256 // N / 4
+  #define N_4_2 512 // 2N / 4
+  #define N_4_3 768 // 3N / 4
+	#define START_Ns 4
+#elif FFT_SIZE == 2048
+  #define N_4_1 512
+  #define N_4_2 1024
+  #define N_4_3 1536
+	#define START_Ns 8
+#endif
+
+#ifdef IFFT
+  #define INV -1.0
+#else
+  #define INV 1.0
+#endif
+
+const float NORM = 1.0 / float(FFT_SIZE);
+  
+shared vec2 row[FFT_SIZE];
+
+vec2 cmul(in vec2 A, in vec2 B) {
+	return vec2(A.x * B.x - A.y * B.y, dot(A, B.yx));
+}
+
+// to prevent shared bank conflicts
+uint swizzle(uint x) {
+  return (x & ~31) | ((x & 31) ^ ((x >> 5) & 31));
+}
+
+uint digit_reverse_4(uint x) {
+  return ((x & 3) << 8) | ((x & 12) << 4) | (x & 48) | ((x & 192) >> 4) | ((x >> 8) & 3);
+}
+
+// in-place Cooley-Tukey FFT, forward if inv = 1, inverse if inv = -1
+void fftR4_CT(uint x, uint Ns) {
+  uint Ns4 = Ns >> 2u; // Ns / 4
+  uint offset = x & (Ns4 - 1u); // x % Ns4;
+
+  // replace with shared memory LUT
+  float angle = -INV * TAU * (float(offset) / float(Ns));
+  vec2 w1 = vec2(cos(angle), sin(angle));
+
+  uint base = (x / Ns4) * Ns + offset; // floor(x / Ns4) * Ns + offset;
+  uint i0 = swizzle(base);
+  uint i1 = swizzle(base + Ns4);
+  uint i2 = swizzle(base + Ns4 * 2u);
+  uint i3 = swizzle(base + Ns4 * 3u);
+
+  vec2 t0 = row[i0];
+  vec2 t1 = cmul(w1, row[i1]);
+  vec2 w2 = cmul(w1, w1);
+  vec2 t2 = cmul(w2, row[i2]);
+  vec2 t3 = cmul(cmul(w2, w1), row[i3]);
+
+  row[i0] = t0 + t1 + t2 + t3;
+  row[i1] = t0 - t2 + vec2(t1.y - t3.y, t3.x - t1.x) * INV;
+  row[i2] = t0 - t1 + t2 - t3;
+  row[i3] = t0 - t2 + vec2(t3.y - t1.y, t1.x - t3.x) * INV;
+}
+
+/* usage:
+  uint y  = gl_GlobalInvocationID.y;
+  uint x0 = gl_GlobalInvocationID.x;
+  uint x1 = x0 + N_4_1;
+  uint x2 = x1 + N_4_1;
+  uint x3 = x2 + N_4_1;
+
+  vec4 t0 = imageLoad(input, ivec2(x0, y), 0);
+  vec4 t1 = imageLoad(input, ivec2(x1, y), 0);
+  vec4 t2 = imageLoad(input, ivec2(x2, y), 0);
+  vec4 t3 = imageLoad(input, ivec2(x3, y), 0);
+
+  rgbaRowFFT(y, x0...x3, t0...t3);
+
+  #if INV == 1
+    float normalizer = 1.0;
+  #else
+    float normalizer = 1.0 / float(N);
+  #endif
+  imageStore(output, ivec2(y, x0), t0 * normalizer);
+  imageStore(output, ivec2(y, x1), t1 * normalizer);
+  imageStore(output, ivec2(y, x2), t2 * normalizer);
+  imageStore(output, ivec2(y, x3), t3 * normalizer);
+*/
+
+// avoid passing x0-x3 to reduce register pressure, only pass x and compute others inside the function, can #define multiples of N_4_1
+void rgbaRowFFT(uint y, uint x, inout vec4 t0, inout vec4 t1, inout vec4 t2, inout vec4 t3) {
+	#if FFT_SIZE == 2048
+		// load rg as real and imag and do radix 2 prepass
+		uint p00 = digit_reverse_4(x) << 1;
+		row[swizzle(p00)]     = t0.rg + t2.rg;
+		row[swizzle(p00 + 1)] = t0.rg - t2.rg;
+
+		uint p01 = digit_reverse_4(x + N_4_1) << 1;
+		row[swizzle(p01)]     = t1.rg + t3.rg;
+		row[swizzle(p01 + 1)] = t1.rg - t3.rg;
+	#else
+		// load rg as real and imag
+		row[swizzle(digit_reverse_4(x))] = t0.rg;
+		row[swizzle(digit_reverse_4(x + N_4_1))] = t1.rg;
+		row[swizzle(digit_reverse_4(x + N_4_2))] = t2.rg;
+		row[swizzle(digit_reverse_4(x + N_4_3))] = t3.rg;
+	#endif
+  barrier();
+
+  for (uint Ns = START_Ns; Ns <= FFT_SIZE; Ns <<= 2) {
+    fftR4_CT(x, Ns);
+    barrier();
+  }
+
+  // store rg fft output
+  t0.rg = row[swizzle(x)];
+  t1.rg = row[swizzle(x + N_4_1)];
+  t2.rg = row[swizzle(x + N_4_2)];
+  t3.rg = row[swizzle(x + N_4_3)];
+  barrier();
+
+	#if FFT_SIZE == 2048
+		// load rg as real and imag and do radix 2 prepass
+		uint p10 = digit_reverse_4(x) << 1;
+		row[swizzle(p10)]     = t0.ba + t2.ba;
+		row[swizzle(p10 + 1)] = t0.ba - t2.ba;
+
+		uint p11 = digit_reverse_4(x + N_4_1) << 1;
+		row[swizzle(p11)]     = t1.ba + t3.ba;
+		row[swizzle(p11 + 1)] = t1.ba - t3.ba;
+	#else
+    // load ba as real and imag
+		row[swizzle(digit_reverse_4(x))] = t0.ba;
+		row[swizzle(digit_reverse_4(x + N_4_1))] = t1.ba;
+		row[swizzle(digit_reverse_4(x + N_4_2))] = t2.ba;
+		row[swizzle(digit_reverse_4(x + N_4_3))] = t3.ba;
+	#endif
+  barrier();
+
+  for (uint Ns = START_Ns; Ns <= FFT_SIZE; Ns <<= 2) {
+    fftR4_CT(x, Ns);
+    barrier();
+  }
+
+  // store combination of r and g FFT output (transposed)
+  t0.ba = row[swizzle(x)];
+  t1.ba = row[swizzle(x + N_4_1)];
+  t2.ba = row[swizzle(x + N_4_2)];
+  t3.ba = row[swizzle(x + N_4_3)];
+}
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/BasicFunctions/PrecomputedAtmosphere.glsl 0.8.22-patched/shaders/Lib/BasicFunctions/PrecomputedAtmosphere.glsl
--- 0.8.22-orig/shaders/Lib/BasicFunctions/PrecomputedAtmosphere.glsl	2026-05-10 15:41:03.585629300 -0700
+++ 0.8.22-patched/shaders/Lib/BasicFunctions/PrecomputedAtmosphere.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -61,14 +61,66 @@
 	return disc;
 }
 
-vec3 RenderMoonDisc(vec3 worldDir, vec3 moonDir){
-	const float angularRadius = SUN_ANGULAR_RADIUS / 180.0 * PI;
+#ifdef MOON_PHASE_LIGHTING
+	vec3 RenderMoonDisc(vec3 worldDir, vec3 moonDir){
+		const float angularRadius = SUN_ANGULAR_RADIUS * PI / 180.0;
 
-	float angle = facos(saturate(dot(worldDir, moonDir)));
+		float cosAngle = dot(worldDir, moonDir);
+		float angle = facos(clamp(cosAngle, -1.0, 1.0));
 
-	float disc = float(angle <= angularRadius);
-	return (vec3(0.8886, 1.0019, 1.3095) * NIGHT_BRIGHTNESS) * disc;
-}
+		// Disc mask (soft edge)
+		float disc = float(angle <= angularRadius);
+
+		float phaseAngle = float(4 + moonPhase) * (-PI / 4.0);
+
+
+		vec3 tangent = normalize(cross(moonDir, vec3(0.0, 1.0, 0.0)));
+		vec3 binormal = cross(moonDir, tangent);
+		vec3 dirDiff = worldDir - moonDir;
+
+		float u = dot(dirDiff, tangent);
+		float v = dot(dirDiff, binormal);
+
+		u = u / (2.0 * angularRadius) + 0.5;
+		v = -v / (2.0 * angularRadius) + 0.5;
+		vec2 moonUV = vec2(u, v);
+
+		vec2 pos = 2.0 * moonUV - 1.0;
+		float r2 = dot(pos, pos); // we got 'em r2
+
+		vec3 moonDirLocal = normalize(vec3(pos.x, pos.y, sqrt(1.0 - r2)));
+
+		vec3 moonDirWorld = moonDirLocal.x * tangent - moonDirLocal.y * binormal - moonDirLocal.z * moonDir;
+
+		vec3 phaseSunDir = normalize(moonDir * cos(phaseAngle) + binormal * sin(phaseAngle));
+
+		float terminator = dot(normalize(moonDirWorld), phaseSunDir);
+
+		// Soft sunlit side
+		float phaseLight = smoothstep(-0.1, 0.05, terminator);
+
+		// Earthshine (dark side glow)
+		// float earthshine = smoothstep(-0.6, -0.1, terminator) * 0.08;
+
+		vec3 moonColor = vec3(0.8886, 1.0019, 1.3095);
+		vec3 earthshineColor = vec3(0.6, 0.7, 1.0) * 0.01; // subtle blue
+
+		vec3 color =
+				moonColor * phaseLight +
+				earthshineColor * (1 - phaseLight);
+
+		return color * disc * (PHASED_NIGHT_BRIGHTNESS + mix(0, 0.00001, float(moonPhase == 4)));
+	}
+#else
+	vec3 RenderMoonDisc(vec3 worldDir, vec3 moonDir){
+		const float angularRadius = SUN_ANGULAR_RADIUS / 180.0 * PI;
+
+		float angle = facos(saturate(dot(worldDir, moonDir)));
+
+		float disc = float(angle <= angularRadius);
+		return (vec3(0.8886, 1.0019, 1.3095) * NIGHT_BRIGHTNESS) * disc;
+	}
+#endif
 
 
 //////////Intersections///////////////////////////
@@ -311,7 +363,7 @@
 		float sun_mu_s = dot(point, sun_direction) / r;
 		float moon_mu_s = dot(point, moon_direction) / r;
 
-		moon_irradiance = (GetTransmittanceToSun(r, moon_mu_s) * (atmosphereModel_solar_irradiance * SUNLIGHT_INTENSITY * NIGHT_BRIGHTNESS)) * LMS;
+		moon_irradiance = (GetTransmittanceToSun(r, moon_mu_s) * (atmosphereModel_solar_irradiance * SUNLIGHT_INTENSITY * PHASED_NIGHT_BRIGHTNESS)) * LMS;
 		return (GetTransmittanceToSun(r, sun_mu_s) * (atmosphereModel_solar_irradiance * SUNLIGHT_INTENSITY)) * LMS;
 }
 
@@ -328,9 +380,9 @@
 		float moon_mu_s = dot(point, moon_direction) / r;
 
 		sun_sky_irradiance = GetIrradiance(r, sun_mu_s) * LMS;
-		moon_sky_irradiance = (GetIrradiance(r, moon_mu_s)) * NIGHT_BRIGHTNESS * LMS;
+		moon_sky_irradiance = (GetIrradiance(r, moon_mu_s)) * PHASED_NIGHT_BRIGHTNESS * LMS;
 
-		moon_irradiance = (GetTransmittanceToSun(r, moon_mu_s) * (atmosphereModel_solar_irradiance * SUNLIGHT_INTENSITY * NIGHT_BRIGHTNESS)) * LMS;
+		moon_irradiance = (GetTransmittanceToSun(r, moon_mu_s) * (atmosphereModel_solar_irradiance * SUNLIGHT_INTENSITY * PHASED_NIGHT_BRIGHTNESS)) * LMS;
 		return (GetTransmittanceToSun(r, sun_mu_s) * (atmosphereModel_solar_irradiance * SUNLIGHT_INTENSITY)) * LMS;
 
 }
@@ -413,7 +465,7 @@
 			float moon_mu_s = dot(planet_surface, moon_direction) / r;
 
 			vec3 sky_irradiance = GetIrradiance(r, sun_mu_s);
-			sky_irradiance += GetIrradiance(r, moon_mu_s) * NIGHT_BRIGHTNESS;
+			sky_irradiance += GetIrradiance(r, moon_mu_s) * PHASED_NIGHT_BRIGHTNESS;
 			vec3 sun_irradiance = atmosphereModel_solar_irradiance * GetTransmittanceToSun(r, sun_mu_s);
 
 			float d = distance(camera, planet_surface);
@@ -425,10 +477,10 @@
 
 
 		vec3 rayleigh = sun_scattering * RayleighPhaseFunction(sun_nu)
-					 + moon_scattering * RayleighPhaseFunction(moon_nu) * NIGHT_BRIGHTNESS;
+					 + moon_scattering * RayleighPhaseFunction(moon_nu) * PHASED_NIGHT_BRIGHTNESS;
 
 		vec3 mie = sun_single_mie_scattering * MiePhaseFunction(atmosphereModel_mie_phase_function_g, sun_nu)
-				+ moon_single_mie_scattering * MiePhaseFunction(atmosphereModel_mie_phase_function_g, moon_nu) * NIGHT_BRIGHTNESS;
+				+ moon_single_mie_scattering * MiePhaseFunction(atmosphereModel_mie_phase_function_g, moon_nu) * PHASED_NIGHT_BRIGHTNESS;
 
 		rayleigh = mix(rayleigh,  vec3(Luminance(rayleigh)) * atmosphereModel_solar_irradiance, wetness * 0.5);
 
@@ -493,10 +545,10 @@
 		moon_single_mie_scattering = moon_single_mie_scattering * smoothstep(0.0, 0.01, moon_mu_s);
 
 		vec3 rayleigh = sun_scattering * RayleighPhaseFunction(sun_nu)
-					 + moon_scattering * RayleighPhaseFunction(moon_nu) * NIGHT_BRIGHTNESS;
+					 + moon_scattering * RayleighPhaseFunction(moon_nu) * PHASED_NIGHT_BRIGHTNESS;
 
 		vec3 mie = sun_single_mie_scattering * MiePhaseFunction(atmosphereModel_mie_phase_function_g, sun_nu)
-				+ moon_single_mie_scattering * MiePhaseFunction(atmosphereModel_mie_phase_function_g, moon_nu) * NIGHT_BRIGHTNESS;
+				+ moon_single_mie_scattering * MiePhaseFunction(atmosphereModel_mie_phase_function_g, moon_nu) * PHASED_NIGHT_BRIGHTNESS;
 
 		rayleigh = mix(rayleigh, vec3(Luminance(rayleigh)) * atmosphereModel_solar_irradiance, wetness * 0.5);
 
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/BasicFunctions/Sunlight_Shadow.glsl 0.8.22-patched/shaders/Lib/BasicFunctions/Sunlight_Shadow.glsl
--- 0.8.22-orig/shaders/Lib/BasicFunctions/Sunlight_Shadow.glsl	2026-05-10 15:41:03.601604100 -0700
+++ 0.8.22-patched/shaders/Lib/BasicFunctions/Sunlight_Shadow.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -155,7 +155,7 @@
 
 			spread = max3(
 				avgDiff * 1e-3 * spread,
-				(0.1 * maskPlant) * shadowProjection[0][0],
+				(0.1 * LEAVES_SSS_MULT * maskPlant) * shadowProjection[0][0],
 				(SHADOW_BASIC_BLUR * 2e-6) / (warpPixelSizeMinMax.x + 2e-3)
 			);
 
@@ -236,7 +236,6 @@
 			result = mix(result, caustics.rgb, caustics.a * rSteps);
 		}
 	}
-	
 	return result;
 }
 
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/IndividualFunctions/DOF.glsl 0.8.22-patched/shaders/Lib/IndividualFunctions/DOF.glsl
--- 0.8.22-orig/shaders/Lib/IndividualFunctions/DOF.glsl	2026-05-10 15:41:03.843377400 -0700
+++ 0.8.22-patched/shaders/Lib/IndividualFunctions/DOF.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -6,9 +6,9 @@
 
 float GetLinearDepth(vec2 coord){
 	#if defined SUPER_RESOLUTION || defined CUSTOM_RENDER_RESOLUTION
-		float depth = textureLod(depthtex0, coord * fsrRenderScale, 0.0).x;
+		float depth = textureLod(depthtex1, coord * fsrRenderScale, 0.0).x;
 	#else
-		float depth = textureLod(depthtex0, coord, 0.0).x;
+		float depth = textureLod(depthtex1, coord, 0.0).x;
 	#endif
 
 	#ifdef LOD_RENDERING
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/IndividualFunctions/NUBIS.glsl 0.8.22-patched/shaders/Lib/IndividualFunctions/NUBIS.glsl
--- 0.8.22-orig/shaders/Lib/IndividualFunctions/NUBIS.glsl	2026-05-10 15:41:03.889973900 -0700
+++ 0.8.22-patched/shaders/Lib/IndividualFunctions/NUBIS.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -16,10 +16,9 @@
 	//Coverage
 		float coverageNoise = textureLod(CloudNoise3D, cloudPos.xyz * vec3(8.5e-5, 1e-5, 8.5e-5) + windDirection + CLOUD_COVERAGE_NOISE_OFFSET, 0.0).x;
 
-		float coverage = mix(CLOUD_CLEAR_COVERY, CLOUD_RAIN_COVERY, wetness);
-		coverage = remapSaturate(1.0 - coverageNoise, coverage * 0.2, coverage);
+		float coverage = remapSaturate(1.0 - coverageNoise, FINAL_CLOUD_COVERAGE * 0.2, FINAL_CLOUD_COVERAGE);
 
-		baseDensity = remapSaturate(baseDensity, coverage, 1.0); 
+		baseDensity = (FINAL_CLOUD_COVERAGE == 0) ? 0 : remapSaturate(baseDensity, coverage, 1.0);
 
 	//Density Control
 		baseDensity *= curve(saturate(cloudPos.w * 1.8 - 0.8)) * 2.0 + 1.0;
@@ -91,7 +90,7 @@
 
 
 
-void NubisCumulus(inout vec3 color, vec3 worldDir, vec2 cloudAltitude, vec3 colorSun, vec3 colorSky, vec3 windDirection, vec3 camera, vec2 noise, out float cloudTransmittance){
+void NubisCumulus(inout vec3 color, vec3 worldDir, vec2 cloudAltitude, vec3 colorSun, vec3 colorSky, vec3 windDirection, vec3 camera, vec2 noise, float maxDepth, out float cloudTransmittance){
 	const float planetRadius = atmosphereModel_bottom_radius * 1e3;
 
 	//Marching Setup
@@ -137,8 +136,11 @@
 		cloudTransmittance = 1.0;
 		vec3 rayHitPos = vec3(0.0);
 		float sumTransmit = 0.0;
+		float maxDepthSq = maxDepth * maxDepth;
+		for (int i = 0; i < int(marchingSteps); i++, marchingPos += marchingIncrement){
+			vec3 toMarchingPos = marchingPos - cameraPosition;
+			if (dot(toMarchingPos, toMarchingPos) > maxDepthSq) break;
 
-		for (int i = 0; i < int(marchingSteps); i++, marchingPos += marchingIncrement){	
 			vec4 cloudPos = SetCloudPos(marchingPos, cloudAltitude, planetRadius, cloudScale, windDirection);
 
 			vec2 cloudDensity = SampleDensity(cloudPos, windDirection, true);
@@ -196,16 +198,31 @@
 }
 
 
-float CloudShadowTex(vec2 coord, vec2 cloudAltitude, vec3 windDirection){
-	coord = coord * CLOUD_SHADOW_RANGE - (CLOUD_SHADOW_RANGE * 0.5);
-	vec4 checkPos = vec4(coord.x + cameraPosition.x, 0.0, coord.y + cameraPosition.z, 0.0);
+float CloudShadowTex(vec3 worldPos){
+
+	vec2 cloudAltitude = vec2(mix(CLOUD_CLEAR_ALTITUDE, CLOUD_RAIN_ALTITUDE, wetness));
+	cloudAltitude.y += mix(CLOUD_CLEAR_THICKNESS, CLOUD_RAIN_THICKNESS, wetness);
 	float thickness = cloudAltitude.y - cloudAltitude.x;
+	if (worldPos.y + cameraPosition.y > cloudAltitude.x) return 1.0;
+	float wind = 0.0005 * (windTime * CLOUD_SPEED + 10.0 * FTC_OFFSET);
+	vec3 windDirection = vec3(1.0, -0.05, -0.4) * wind;
+	#ifdef NO_COMPOSITE_VS
+		vec3 worldShadowVector = shadowModelViewInverse2;
+	#endif
+	vec2 coord = worldPos.xz - worldShadowVector.xz * ((worldPos.y + cameraPosition.y) / worldShadowVector.y);
+
+	vec4 checkPos = vec4(coord.x + cameraPosition.x, 0.0, coord.y + cameraPosition.z, 0.0);
 
 	checkPos += vec4(worldShadowVector * ((cloudAltitude.x + thickness * 0.25) / worldShadowVector.y) , 0.25);
 	float cloudDensity = SampleDensity(checkPos, windDirection, false).y;
-	
+
 	checkPos += vec4(worldShadowVector * ((thickness * 0.25) / worldShadowVector.y) , 0.25);
 	cloudDensity = max(cloudDensity, SampleDensity(checkPos, windDirection, false).y);
 
-	return exp2(-cloudDensity * 20.0);
+	float cloudShadowDamping = smoothstep(0, 0.02, .25 - abs(shadowAngle - 0.25));
+
+	// return mix(1.0, exp2(-cloudDensity * 20.0), cloudShadowDamping);
+	// return exp2(-cloudDensity * 20.0);
+	// return exp2(-smoothstep(0,1,cloudDensity) * 20.0);
+	return mix(1 - FINAL_CLOUD_COVERAGE, exp2(-smoothstep(0,1,cloudDensity) * 20.0), cloudShadowDamping);
 }
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/IndividualFunctions/PlanarClouds.glsl 0.8.22-patched/shaders/Lib/IndividualFunctions/PlanarClouds.glsl
--- 0.8.22-orig/shaders/Lib/IndividualFunctions/PlanarClouds.glsl	2026-05-10 15:41:03.923129700 -0700
+++ 0.8.22-patched/shaders/Lib/IndividualFunctions/PlanarClouds.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -110,10 +110,10 @@
 	float intersection = RaySphereIntersection(rayStartPos, worldDir, planetRadius + PC_ALTITUDE).y;
 
 	if (intersection > 0.0){
-		float wind = 2.5 * (frameTimeCounter * CLOUD_SPEED + 10.0 * FTC_OFFSET);
+		float wind = 2.5 * (windTime * CLOUD_SPEED + 10.0 * FTC_OFFSET);
 		vec2 windDirection = vec2(-wind, wind * 0.2);
 
-		vec2 coverageDensity = mix(vec2(PC_CLEAR_COVERAGE, PC_CLEAR_DENSITY), vec2(PC_RAIN_COVERAGE, PC_RAIN_DENSITY), wetness);
+		vec2 coverageDensity = mix(vec2(FINAL_PC_COVERAGE, PC_CLEAR_DENSITY), vec2(PC_RAIN_COVERAGE, PC_RAIN_DENSITY), wetness);
 
 		float VdotL = dot(worldDir, worldShadowVector) * 0.5 + 0.5;
 
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/IndividualFunctions/VolumetricFog.glsl 0.8.22-patched/shaders/Lib/IndividualFunctions/VolumetricFog.glsl
--- 0.8.22-orig/shaders/Lib/IndividualFunctions/VolumetricFog.glsl	2026-05-10 15:41:03.970044100 -0700
+++ 0.8.22-patched/shaders/Lib/IndividualFunctions/VolumetricFog.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -1,4 +1,6 @@
-
+#ifdef CLOUD_SHADOW
+	#include "/Lib/IndividualFunctions/NUBIS.glsl"
+#endif
 
 #if VFOG_NOISE_TYPE == 0
 
@@ -31,10 +33,13 @@
 
 float FogDensity(vec3 fogPos, float baseDensity, float fogTimeFactor, vec3 wind){
 	const float maxHeight = max(VFOG_HEIGHT, VFOG_HEIGHT_2);
+	const float minHeight = min(VFOG_HEIGHT, VFOG_HEIGHT_2);
 
-	float dh = abs(fogPos.y - maxHeight);
-	dh *= (2.5 / VFOG_FALLOFF) * (1.0 - fogTime.y * 0.5);
-	float falloff = exp2(-dh);
+	vec2 dh = vec2(abs(fogPos.y - maxHeight), max(fogPos.y - maxHeight, 0.0) + max(minHeight - fogPos.y, 0.0));
+	dh *= (vec2(2.5, 2.0) / VFOG_FALLOFF) * (1.0 - fogTime.y * 0.5);
+	vec2 falloff2 = exp2(-dh);
+	float falloff = falloff2.x;
+	float heightDensity = falloff2.y * .15;
  
 	
 	fogPos = fogPos * vec3(0.2, 0.05, 0.2) + wind;
@@ -60,7 +65,7 @@
 	//density  = density * density;
 	density *= density * falloff;
 
-	density = (density * VFOG_DENSITY + baseDensity) * fogTimeFactor;
+	density = ((heightDensity + density) * VFOG_DENSITY + baseDensity) * fogTimeFactor;
 
 	return density;
 }
@@ -68,9 +73,9 @@
 #endif
 
 void VolumetricFog(inout vec3 color, vec3 startPos, vec3 endPos, vec3 worldDir, float globalCloudShadow, float fogTimeFactor){
-	#if VFOG_NOISE_TYPE == 0
+	//#if VFOG_NOISE_TYPE == 0
 		fogTimeFactor *= fogTime.y * 2.0 + 1.0;
-	#endif
+	//#endif
 
 	#if defined TAA || defined SUPER_RESOLUTION
 		float noise = BlueNoiseTemporal().y;
@@ -112,7 +117,7 @@
 	float phases = HenyeyGreenstein(VdotL, 0.65) * 0.7 + 0.035;
 
 	#if VFOG_NOISE_TYPE > 0
-		float windTimer = (frameTimeCounter * CLOUD_SPEED + 10.0 * FTC_OFFSET);
+		float windTimer = (windTime * CLOUD_SPEED + 10.0 * FTC_OFFSET);
 		vec3 wind = (vec3(0.6, -0.05, -1.0) * 0.1) * windTimer;
 	#endif
 
@@ -186,25 +191,24 @@
 		#endif
 		}
 
-/*
 		#ifdef VOLUMETRIC_CLOUDS
 			#ifdef CLOUD_SHADOW
 				#ifdef VFOG_CLOUD_SHADOW
-					float cloudShadow = CloudShadowFromTex(fogPos - cameraPosition);
+					float cloudShadow = CloudShadowTex(fogPos - cameraPosition);
 				#endif
 			#endif
 		#endif
-*/
+
 		
 		float integral = transmittance - absorption * transmittance;
 		stepRayDensity *= integral;
 
-		//#if defined VOLUMETRIC_CLOUDS && defined CLOUD_SHADOW && defined VFOG_CLOUD_SHADOW
-		//	stepRayDensity *= mix(cloudShadow, 1.0, wetness * 0.1 + 0.03);
-		//	fogDensity += integral * mix(cloudShadow, 1.0, wetness * 0.3 + 0.7);
-		//#else
+		#if defined VOLUMETRIC_CLOUDS && defined CLOUD_SHADOW && defined VFOG_CLOUD_SHADOW
+			stepRayDensity *= mix(cloudShadow, 1.0, wetness * 0.1 + 0.03);
+			fogDensity += integral * mix(cloudShadow, 1.0, wetness * 0.3 + 0.7);
+		#else
 			fogDensity += integral;
-		//#endif
+		#endif
 
 		rayDensity += stepRayDensity * shadow;
 
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/PathTracing/Tracer/SpecularTracer.glsl 0.8.22-patched/shaders/Lib/PathTracing/Tracer/SpecularTracer.glsl
--- 0.8.22-orig/shaders/Lib/PathTracing/Tracer/SpecularTracer.glsl	2026-05-10 15:41:04.184803100 -0700
+++ 0.8.22-patched/shaders/Lib/PathTracing/Tracer/SpecularTracer.glsl	2026-05-23 23:26:31.777502900 -0700
@@ -335,7 +335,10 @@
 
 					#ifdef DIMENSION_OVERWORLD
 					#ifdef SUNLIGHT_LEAK_FIX
-						sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+						// sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+						float originalMask = saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+						float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+						sunLighting *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 					#endif
 					#endif
 					/*
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/DiffuseTracing_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/DiffuseTracing_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/DiffuseTracing_FS.glsl	2026-05-10 15:41:04.640234000 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/DiffuseTracing_FS.glsl	2026-05-23 23:26:41.023811300 -0700
@@ -52,6 +52,82 @@
 	#include "/Lib/IndividualFunctions/EndSkyTimer.glsl"
 #endif
 
+#ifdef CLOUD_SHADOW
+vec2 SampleDensity(vec4 cloudPos, vec3 windDirection, bool sampleDetail){
+	//Base Noise
+		vec4 baseNoise = textureLod(CloudNoise3D, cloudPos.xyz * CLOUD_BASE_NOISE_SCALE + windDirection * 10.0, 0.0);
+
+		float baseDensity = baseNoise.y * 0.4 + baseNoise.z * 0.4 + baseNoise.w * 0.2;
+		baseDensity = remapSaturate(baseNoise.x, baseDensity - 1.0, 1.0);
+
+		float shapeCurve = curveTop(saturate(1.0 - cloudPos.w)) * 0.5;
+		shapeCurve += curve(saturate(1.15 - cloudPos.w * 1.43)) * 0.5;
+		baseDensity *= shapeCurve;
+		baseDensity *= fsqrt(saturate(cloudPos.w * 2.5)) * 0.2 + 0.8;
+		baseDensity *= mix(1.0, curveTop(saturate(cloudPos.w * 3.0)), wetness);
+
+	//Coverage
+		float coverageNoise = textureLod(CloudNoise3D, cloudPos.xyz * vec3(8.5e-5, 1e-5, 8.5e-5) + windDirection + CLOUD_COVERAGE_NOISE_OFFSET, 0.0).x;
+
+		float coverage = remapSaturate(1.0 - coverageNoise, FINAL_CLOUD_COVERAGE * 0.2, FINAL_CLOUD_COVERAGE);
+
+		baseDensity = (FINAL_CLOUD_COVERAGE == 0) ? 0 : remapSaturate(baseDensity, coverage, 1.0);
+
+	//Density Control
+		baseDensity *= curve(saturate(cloudPos.w * 1.8 - 0.8)) * 2.0 + 1.0;
+		baseDensity *= curveTop(saturate(cloudPos.w * 1.8));
+
+		baseDensity = saturate(baseDensity);
+
+
+	float detailedDensity = 0.0;
+
+	if(baseDensity > 1e-6 && sampleDetail){
+		//Curl Noise
+			float curlNoise = textureLod(noisetex, cloudPos.xy * 0.0001, 0.0).w;
+			cloudPos.xy += curlNoise * (100.0 - cloudPos.w * 100.0);
+
+		//Detailed Noise
+			vec3 detailedNoise = textureLod(CloudDetailedNoise3D, cloudPos.xyz * CLOUD_DETAILED_NOISE_SCALE + windDirection * 140.0, 0.0).xyz;
+			detailedDensity = 1.0 - detailedNoise.x * 0.625 - detailedNoise.y * 0.25 - detailedNoise.z * 0.125;
+			detailedDensity *= (0.21 - wetness * 0.11) * CLOUD_DETAILED_NOISE_STRENGTH;
+
+			detailedDensity = remapSaturate(baseDensity, detailedDensity, 1.0);
+	}else{
+		detailedDensity = baseDensity;
+	}
+
+	return vec2(baseDensity, detailedDensity);
+}
+
+float CloudShadowTex(vec3 worldPos){
+
+	vec2 cloudAltitude = vec2(mix(CLOUD_CLEAR_ALTITUDE, CLOUD_RAIN_ALTITUDE, wetness));
+	cloudAltitude.y += mix(CLOUD_CLEAR_THICKNESS, CLOUD_RAIN_THICKNESS, wetness);
+	float thickness = cloudAltitude.y - cloudAltitude.x;
+	if (worldPos.y + cameraPosition.y > cloudAltitude.x) return 1.0;
+	float wind = 0.0005 * (windTime * CLOUD_SPEED + 10.0 * FTC_OFFSET);
+	vec3 windDirection = vec3(1.0, -0.05, -0.4) * wind;
+	vec3 worldShadowVector = shadowModelViewInverse2;
+	vec2 coord = worldPos.xz - worldShadowVector.xz * ((worldPos.y + cameraPosition.y) / worldShadowVector.y);
+
+	vec4 checkPos = vec4(coord.x + cameraPosition.x, 0.0, coord.y + cameraPosition.z, 0.0);
+
+	checkPos += vec4(worldShadowVector * ((cloudAltitude.x + thickness * 0.25) / worldShadowVector.y) , 0.25);
+	float cloudDensity = SampleDensity(checkPos, windDirection, false).y;
+
+	checkPos += vec4(worldShadowVector * ((thickness * 0.25) / worldShadowVector.y) , 0.25);
+	cloudDensity = max(cloudDensity, SampleDensity(checkPos, windDirection, false).y);
+
+	float cloudShadowDamping = smoothstep(0, 0.02, .25 - abs(shadowAngle - 0.25));
+
+	// return mix(1.0, exp2(-cloudDensity * 20.0), cloudShadowDamping);
+	// return exp2(-cloudDensity * 20.0);
+	// return exp2(-smoothstep(0,1,cloudDensity) * 20.0);
+	return mix(1 - FINAL_CLOUD_COVERAGE, exp2(-smoothstep(0,1,cloudDensity) * 20.0), cloudShadowDamping);
+}
+#endif
+
 
 vec3 SkyLighting(vec3 raydir, float pdf, float lightmap){
 	//float SdotN = dot(worldNormal, normalize(worldSunVector + vec3(0.0, 1.0, 0.0)));
@@ -508,7 +584,10 @@
 					#endif
 					#ifdef DIMENSION_OVERWORLD
 					#ifdef SUNLIGHT_LEAK_FIX
-						sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+						// sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+						float originalMask = saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+						float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+						sunLighting *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 					#endif
 					#endif
 					/*
@@ -524,6 +603,10 @@
 					if (sunLighting > 0.0){
 						vec3 hitWorldPos = hitVoxelPos - cameraPositionFract - (voxelResolution * 0.5);
 						vec3 sunColor = sunLight * SimpleShadow(hitWorldPos, hitNormal) * sunLighting;
+						
+						#ifdef CLOUD_SHADOW
+							sunColor *= CloudShadowTex(worldPos);
+						#endif
 						diffuse += sunColor * hitSurface;
 					}
 				#endif
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/IRC_CS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/IRC_CS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/IRC_CS.glsl	2026-05-10 15:41:04.721176400 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/IRC_CS.glsl	2026-05-23 23:26:51.492316200 -0700
@@ -480,7 +480,10 @@
 						#endif
 						#ifdef DIMENSION_OVERWORLD
 						#ifdef SUNLIGHT_LEAK_FIX
-							sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+							// sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+							float originalMask = saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+							float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+							sunLighting *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 						#endif
 						#endif
 						/*
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl	2026-05-26 19:07:25.859702700 -0700
@@ -0,0 +1,105 @@
+// by huj31415
+
+// unpack image and kernel spectra and complex multiply
+// ifft and save to image
+
+#include "/Lib/UniformDeclare.glsl"
+#include "/Lib/Utilities.glsl"
+
+#include "/Lib/BasicFunctions/FFT.glsl"
+
+// N rows, N/4 threads per row
+const ivec3 workGroups = ivec3(1, FFT_SIZE, 1);
+layout (local_size_x = N_4_1, local_size_y = 1) in;
+
+#define IFFT
+
+#ifdef ROW
+// Row pass ===================================================================
+
+layout (rgba32f) uniform image2D kerFFT;
+layout (rgba32f) uniform image2D imgFFT;
+layout (rgba32f) uniform writeonly image2D tmpFFT;
+
+vec4 freqConvolve(ivec2 coord) {
+  if (coord.x >= FFT_SIZE || coord.y >= FFT_SIZE) return vec4(0.0);
+
+  // get symmetric coordinate -k = (N - x) % N
+  ivec2 symCoord = (ivec2(FFT_SIZE) - coord) & ivec2(FFT_SIZE - 1); // Modulo using bitwise AND for power-of-two sizes
+
+  // fetch current and symmetric samples for both image and kernel
+
+  // unpack and convolve, 0.5 factored out and applied at end
+  vec4 h_img    = imageLoad(imgFFT, coord);
+  vec4 h_imgSym = imageLoad(imgFFT, symCoord);
+  vec4 h_ker    = imageLoad(kerFFT, coord);
+  vec4 h_kerSym = imageLoad(kerFFT, symCoord);
+  
+  vec2 F_R = vec2(h_img.x + h_imgSym.x, h_img.y - h_imgSym.y);
+  vec2 K_R = vec2(h_ker.x + h_kerSym.x, h_ker.y - h_kerSym.y);
+  vec2 Out_R = cmul(F_R, K_R);
+
+  vec2 F_G = vec2(h_img.y + h_imgSym.y, h_imgSym.x - h_img.x);
+  vec2 K_G = vec2(h_ker.y + h_kerSym.y, h_kerSym.x - h_ker.x);
+  vec2 Out_G = cmul(F_G, K_G);
+
+  vec2 F_B = vec2(h_img.z + h_imgSym.z, h_img.w - h_imgSym.w);
+  vec2 K_B = vec2(h_ker.z + h_kerSym.z, h_ker.y - h_kerSym.y);
+  vec2 Out_B = cmul(F_B, K_B);
+
+  vec2 F_A = vec2(h_img.w + h_imgSym.w, h_imgSym.z - h_img.z);
+  vec2 K_A = vec2(h_ker.w + h_kerSym.w, h_kerSym.z - h_ker.z);
+  vec2 Out_A = cmul(F_A, K_A);
+
+  // repack into rgba for ifft
+  return 0.25 * vec4(
+    Out_R.x - Out_G.y,
+    Out_R.y + Out_G.x,
+    Out_B.x - Out_A.y,
+    Out_B.y + Out_A.x
+  );
+}
+
+
+
+void main() {
+  uint y = gl_GlobalInvocationID.y;
+  uint x = gl_GlobalInvocationID.x;
+
+  vec4 t0 = freqConvolve(ivec2(x, y));
+  vec4 t1 = freqConvolve(ivec2(x + N_4_1, y));
+  vec4 t2 = freqConvolve(ivec2(x + N_4_2, y));
+  vec4 t3 = freqConvolve(ivec2(x + N_4_3, y));
+
+  rgbaRowFFT(y, x, t0, t1, t2, t3);
+
+  imageStore(tmpFFT, ivec2(y, x), t0);
+  imageStore(tmpFFT, ivec2(y, x + N_4_1), t1);
+  imageStore(tmpFFT, ivec2(y, x + N_4_2), t2);
+  imageStore(tmpFFT, ivec2(y, x + N_4_3), t3);
+}
+
+#else
+// Column pass ================================================================
+
+layout (rgba32f) uniform image2D tmpFFT;
+layout (rgba32f) uniform writeonly image2D imgFFT;
+
+void main() {
+  uint y = gl_GlobalInvocationID.y;
+  uint x = gl_GlobalInvocationID.x;
+
+  vec4 t0 = imageLoad(tmpFFT, ivec2(x, y));
+  vec4 t1 = imageLoad(tmpFFT, ivec2(x + N_4_1, y));
+  vec4 t2 = imageLoad(tmpFFT, ivec2(x + N_4_2, y));
+  vec4 t3 = imageLoad(tmpFFT, ivec2(x + N_4_3, y));
+
+  rgbaRowFFT(y, x, t0, t1, t2, t3);
+
+  imageStore(imgFFT, ivec2(y, x), t0 * NORM);
+  imageStore(imgFFT, ivec2(y, x + N_4_1), t1 * NORM);
+  imageStore(imgFFT, ivec2(y, x + N_4_2), t2 * NORM);
+  imageStore(imgFFT, ivec2(y, x + N_4_3), t3 * NORM);
+}
+
+#endif
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/ImageFFT_CS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/ImageFFT_CS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/ImageFFT_CS.glsl	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/ImageFFT_CS.glsl	2026-05-28 14:31:55.413587600 -0700
@@ -0,0 +1,95 @@
+// by huj31415
+
+// run once per frame
+// compute FFT of image and write to image
+
+#include "/Lib/UniformDeclare.glsl"
+#include "/Lib/Utilities.glsl"
+
+#include "/Lib/BasicFunctions/FFT.glsl"
+
+// N rows, N/4 threads per row
+const ivec3 workGroups = ivec3(1, FFT_SIZE, 1);
+layout (local_size_x = N_4_1, local_size_y = 1) in;
+
+#ifdef ROW
+// Row pass ===================================================================
+
+// FBTEX_MAIN_OUTPUT already declared in UniformDeclare.glsl
+layout (rgba32f) uniform writeonly image2D tmpFFT;
+
+float GetExposureValue(){
+	return texelFetch(pixelData2D, ivec2(PIXELDATA_EXPOSURE, 0), 0).y;
+}
+
+#if CAP_LUMINANCE > -1
+  const float maxLuminance = float(1 << CAP_LUMINANCE);
+#endif
+
+// soft knee thresholding
+vec4 softKnee(vec4 color) {
+  float l = Luminance(color.rgb);
+  float x = max(0, l - BLOOM_THRESHOLD + BLOOM_KNEE);
+  color.rgb *= mix(x * x / (4 * BLOOM_KNEE), max(0, l - BLOOM_THRESHOLD), float(x > BLOOM_THRESHOLD + BLOOM_KNEE)) / max(l, 1e-6);
+  #if CAP_LUMINANCE > -1
+    color *= min(1.0, maxLuminance / l);
+  #endif
+  return color;
+}
+
+// downsample and pad scene texture into the FFT workspace with correct aspect ratio
+vec4 sampleThresholdedSceneTex(uint x, uint y) {
+  float maxDim = float(FFT_SIZE - KERNEL_SIZE);
+  vec2 size = aspectRatio >= 1.0 ? vec2(maxDim, maxDim / aspectRatio) : vec2(maxDim * aspectRatio, maxDim);
+  // zero pad outside valid image region
+  if (any(greaterThanEqual(vec2(x, y), size))) {
+    return vec4(0.0);
+  }
+
+  // map thread coordinates to uv
+  vec2 uv = (vec2(x, y) + 0.5) / size;
+
+  return softKnee(textureLod(FBTEX_MAIN_OUTPUT, uv, 0) * GetExposureValue());
+}
+
+void main() {
+  uint y = gl_GlobalInvocationID.y;
+  uint x = gl_GlobalInvocationID.x;
+
+  vec4 t0 = sampleThresholdedSceneTex(x, y);
+  vec4 t1 = sampleThresholdedSceneTex(x + N_4_1, y);
+  vec4 t2 = sampleThresholdedSceneTex(x + N_4_2, y);
+  vec4 t3 = sampleThresholdedSceneTex(x + N_4_3, y);
+
+  rgbaRowFFT(y, x, t0, t1, t2, t3);
+
+  imageStore(tmpFFT, ivec2(y, x),         t0);
+  imageStore(tmpFFT, ivec2(y, x + N_4_1), t1);
+  imageStore(tmpFFT, ivec2(y, x + N_4_2), t2);
+  imageStore(tmpFFT, ivec2(y, x + N_4_3), t3);
+}
+
+#else
+// Column pass ================================================================
+
+layout (rgba32f) uniform image2D tmpFFT;
+layout (rgba32f) uniform writeonly image2D imgFFT;
+
+void main() {
+  uint y = gl_GlobalInvocationID.y;
+  uint x = gl_GlobalInvocationID.x;
+
+  vec4 t0 = imageLoad(tmpFFT, ivec2(x, y));
+  vec4 t1 = imageLoad(tmpFFT, ivec2(x + N_4_1, y));
+  vec4 t2 = imageLoad(tmpFFT, ivec2(x + N_4_2, y));
+  vec4 t3 = imageLoad(tmpFFT, ivec2(x + N_4_3, y));
+
+  rgbaRowFFT(y, x, t0, t1, t2, t3);
+
+  imageStore(imgFFT, ivec2(y, x),         t0 * NORM);
+  imageStore(imgFFT, ivec2(y, x + N_4_1), t1 * NORM);
+  imageStore(imgFFT, ivec2(y, x + N_4_2), t2 * NORM);
+  imageStore(imgFFT, ivec2(y, x + N_4_3), t3 * NORM);
+}
+
+#endif
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/PixelData_VS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/PixelData_VS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/PixelData_VS.glsl	2026-05-10 15:41:04.773780200 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/PixelData_VS.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -46,11 +46,11 @@
 				if (heldItemId != 11000.0 && centerMaterialID == MATID_HAND){
 					centerDepth = texelFetch(depthtex2, centerTexelCoord, 0).x;
 				}else{
-					centerDepth = texelFetch(depthtex0, centerTexelCoord, 0).x;
+					centerDepth = texelFetch(depthtex1, centerTexelCoord, 0).x;
 				}
 
 			#else
-				float centerDepth = texelFetch(depthtex0, centerTexelCoord, 0).x;
+				float centerDepth = texelFetch(depthtex1, centerTexelCoord, 0).x;
 
 			#endif
 
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/SH_TRACING_CS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/SH_TRACING_CS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/SH_TRACING_CS.glsl	2026-05-10 15:41:04.821135600 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/SH_TRACING_CS.glsl	2026-05-23 23:26:55.340022000 -0700
@@ -270,7 +270,10 @@
 			#endif
 			#ifdef DIMENSION_OVERWORLD
 			#ifdef SUNLIGHT_LEAK_FIX
-				sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+				// sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+				float originalMask = saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+				float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+				sunLighting *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 			#endif
 			#endif
 			/*
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/ShadowSample_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/ShadowSample_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/ShadowSample_FS.glsl	2026-05-10 15:41:04.805379300 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/ShadowSample_FS.glsl	2026-05-23 23:27:01.909028900 -0700
@@ -118,7 +118,10 @@
 		#endif
 		#ifdef DIMENSION_OVERWORLD
 		#ifdef SUNLIGHT_LEAK_FIX
-			float lightMask = saturate(gbuffer.lightmap.g * 1e5 + float(isEyeInWater == 1));
+			// float lightMask = saturate(gbuffer.lightmap.g * 1e5 + float(isEyeInWater == 1));
+			float originalMask = saturate(gbuffer.lightmap.g * 1e5 + float(isEyeInWater == 1));
+			float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+			float lightMask = mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 			shadow *= lightMask;
 		#endif
 		#endif
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/SkyImage_CS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/SkyImage_CS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/SkyImage_CS.glsl	2026-05-10 15:41:04.837623800 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/SkyImage_CS.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -110,7 +110,7 @@
 		float cloudTransmittance = 1.0;
 		#ifdef VOLUMETRIC_CLOUDS
 			if ((cameraPosition.y > cloudAltitude.x || !ray_r_mu_intersects_ground) && (cameraPosition.y < cloudAltitude.y || worldDir.y < 0.0))
-				NubisCumulus(skyImage, worldDir, cloudAltitude, colorShadowlight, colorSkylight, windDirection, camera, noise_0, cloudTransmittance);
+				NubisCumulus(skyImage, worldDir, cloudAltitude, colorShadowlight, colorSkylight, windDirection, camera, noise_0, 1e9, cloudTransmittance);
 		#endif
 
 		#ifdef PLANAR_CLOUDS
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/Sky_Overworld_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/Sky_Overworld_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/Sky_Overworld_FS.glsl	2026-05-10 15:41:04.870457400 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/Sky_Overworld_FS.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -33,11 +33,11 @@
 
 vec3 HashStars(vec3 worldDir){
 	#if FSR2_SCALE >= 0
-		const float scale = 300.0 - float(FSR2_SCALE) * 40.0;
-		const float coverage = 0.007 + FSR2_SCALE * 0.003;
+		const float scale = 500.0 - float(FSR2_SCALE) * 40.0; // 300 -
+		const float coverage = 0.01 + FSR2_SCALE * 0.005; // 0.07 0.03
 	#else
-		const float scale = 384.0;
-		const float coverage = 0.007;
+		const float scale = 512.0; // 384
+		const float coverage = 0.01; // 0.007
 	#endif
 	const float maxLuminance = 0.05 * HASHSTARS_BRIGHTNESS;
 	const float minTemperature = 4000.0;
@@ -57,7 +57,7 @@
 	hash.y = 2.0 * hash.y - 4.0 * hash.y * hash.y + 3.0 * hash.y * hash.y * hash.y;
 
 	float c = remapSaturate(hash.x, 1.0 - coverage, 1.0);
-	return (maxLuminance * remapSaturate(r, 0.25, 0.0) * c * c) * Blackbody(mix(minTemperature, maxTemperature, hash.y));
+	return (maxLuminance * remapSaturate(r, 0.125, 0.0) * c * c) * Blackbody(mix(minTemperature, maxTemperature, hash.y)); // .25
 }
 
 ////////////////////////////// Main //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
@@ -90,9 +90,9 @@
 	vec2 cloudAltitude = vec2(mix(CLOUD_CLEAR_ALTITUDE, CLOUD_RAIN_ALTITUDE, wetness));
 	cloudAltitude.y += mix(CLOUD_CLEAR_THICKNESS, CLOUD_RAIN_THICKNESS, wetness);
 
-	float wind = 0.0005 * (frameTimeCounter * CLOUD_SPEED + 10.0 * FTC_OFFSET);
+	float wind = 0.0005 * (windTime * CLOUD_SPEED + 10.0 * FTC_OFFSET);
 	//wind = 0.0005 * mod(frameTimeCounter * 50.0, 3600.0);
-	vec3 windDirection = vec3(1.0, wetness * 0.1 - 0.05, -0.4) * wind;
+	vec3 windDirection = vec3(1.0, -0.05, -0.4) * wind;
 
 
 //////////////////// Sky ///////////////////////////////////////////////////////////////////////////
@@ -115,7 +115,7 @@
 	float cloudTransmittance = 1.0;
 	#ifdef VOLUMETRIC_CLOUDS
 		if ((cameraPosition.y > cloudAltitude.x || !ray_r_mu_intersects_ground) && (cameraPosition.y < cloudAltitude.y || worldDir.y < 0.0))
-			NubisCumulus(color, worldDir, cloudAltitude, colorShadowlight, colorSkylight, windDirection, camera, noise_1, cloudTransmittance);
+			NubisCumulus(color, worldDir, cloudAltitude, colorShadowlight, colorSkylight, windDirection, camera, noise_1, 1e9, cloudTransmittance);
 	#endif
 
 	#ifdef PLANAR_CLOUDS
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/Soild_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/Soild_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/Soild_FS.glsl	2026-05-10 15:41:04.887489000 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/Soild_FS.glsl	2026-05-23 23:27:09.272842200 -0700
@@ -183,9 +183,14 @@
 	in vec3 shadowVector;
 
 	in vec3 colorShadowlight;
+	
 #endif
 
 
+#if defined CLOUD_SHADOW && defined DIMENSION_OVERWORLD
+	#include "/Lib/IndividualFunctions/NUBIS.glsl"
+#endif
+
 #include "/Lib/GbufferData.glsl"
 #include "/Lib/Uniform/GbufferTransforms.glsl"
 
@@ -556,18 +561,15 @@
 			float shadowmapRange = saturate(3.2 - opaqueDist / min(shadowDistance, far) * 4.0);
 		#endif
 
-	/*
-		#ifdef VOLUMETRIC_CLOUDS
+		#if defined VOLUMETRIC_CLOUDS && defined DIMENSION_OVERWORLD
 			#ifdef CLOUD_SHADOW
-				float cloudShadow = CloudShadowFromTex(worldPos);
+				float cloudShadow = CloudShadowTex(worldPos);
 			#else
 				float cloudShadow = 1.0 - wetness * RAIN_SHADOW;
 			#endif
 		#else
 			float cloudShadow = 1.0 - wetness * RAIN_SHADOW;
 		#endif
-	*/
-		float cloudShadow = 1.0 - wetness * RAIN_SHADOW;
 
 		gbuffer.material.scattering *= saturate(1.0 - materialMask.leaves - materialMask.grass);
 		gbuffer.worldNormal = normalize(mix(gbuffer.worldNormal, vec3(0.0, 1.0, 0.0), materialMask.grass * 0.49));
@@ -621,7 +623,11 @@
 			#endif
 			#ifdef DIMENSION_OVERWORLD
 			#ifdef SUNLIGHT_LEAK_FIX
-				float lightMask = saturate(gbuffer.lightmap.g * 1e5 + float(isEyeInWater == 1));
+				// float lightMask = saturate(gbuffer.lightmap.g * 1e5 + float(isEyeInWater == 1));
+				float originalMask = saturate(gbuffer.lightmap.g * 1e5 + float(isEyeInWater == 1));
+				float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+				float lightMask = mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
+
 				shadow *= lightMask;
 			#endif
 			#endif
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/SpecularTracing_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/SpecularTracing_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/SpecularTracing_FS.glsl	2026-05-10 15:41:04.971673300 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/SpecularTracing_FS.glsl	2026-05-23 23:27:12.674103000 -0700
@@ -92,7 +92,10 @@
 		#ifdef DIMENSION_OVERWORLD
 			float highlightStrength = float(isSmooth && (isEyeInWater == 0 || materialMask.water < 0.5)) * shadowHighlightStrength;
 			#ifdef SUNLIGHT_LEAK_FIX
-				highlightStrength *= saturate(gbuffer.lightmap.g * 1e10);
+				// highlightStrength *= saturate(gbuffer.lightmap.g * 1e10);
+				float originalMask = saturate(gbuffer.lightmap.g * 1e10);
+				float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+				highlightStrength *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 			#endif
 		#else
 			const float highlightStrength = 0.0;
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/Translucent_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/Translucent_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/Translucent_FS.glsl	2026-05-10 15:41:05.022866400 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/Translucent_FS.glsl	2026-05-23 23:27:16.386664500 -0700
@@ -446,7 +446,10 @@
 		#ifdef DIMENSION_OVERWORLD
 			sunlight *= 1.0 - wetness * RAIN_SHADOW;
 			#ifdef SUNLIGHT_LEAK_FIX
-				sunlight *= saturate(lightmap.g * 1e5 + float(isEyeInWater == 1));
+				// sunlight *= saturate(lightmap.g * 1e5 + float(isEyeInWater == 1));
+				float originalMask = saturate(lightmap.g * 1e5 + float(isEyeInWater == 1));
+				float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+				sunlight *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 			#endif
 			particleColor += ParticleShadow(worldPos, voxelPos, particleNormal, waterDist, sunlight, lightmap.g) * colorSunlight;
 		#else
@@ -710,6 +713,9 @@
 						float highlightStrength = shadowHighlightStrength;
 						#ifdef SUNLIGHT_LEAK_FIX
 							highlightStrength *= saturate(gbuffer.lightmap.g * 1e10);
+							// float originalMask = saturate(gbuffer.lightmap.g * 1e10);
+							// float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+							// highlightStrength *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 						#endif
 					#else
 						const float highlightStrength = 0.0;
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Composite/Volumetric_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Composite/Volumetric_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Composite/Volumetric_FS.glsl	2026-05-10 15:41:05.038382300 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Composite/Volumetric_FS.glsl	2026-05-10 19:20:16.000000000 -0700
@@ -58,6 +58,9 @@
 
 #ifdef DIMENSION_OVERWORLD
 	#include "/Lib/IndividualFunctions/VolumetricFog.glsl"
+	#ifndef CLOUD_SHADOW
+		#include "/Lib/IndividualFunctions/NUBIS.glsl"
+	#endif
 	#define FULL_WATERFOG
 	#include "/Lib/IndividualFunctions/WaterFog.glsl"
 #endif
@@ -153,7 +156,27 @@
 		#endif
 
 		#ifdef LANDSCATTERING
-			if (isEyeInWater == 0) LandAtmosphericScattering(color, waterDist, vec3(0.0), rayWorldPos, worldDir, materialMask.sky > 0.5);
+			if (isEyeInWater == 0) {
+				#ifdef VOLUMETRIC_CLOUDS
+					// change order of land scattering and volumetric clouds based on camera position
+					vec2 cloudAltitude = vec2(mix(CLOUD_CLEAR_ALTITUDE, CLOUD_RAIN_ALTITUDE, wetness));
+					cloudAltitude.y += mix(CLOUD_CLEAR_THICKNESS, CLOUD_RAIN_THICKNESS, wetness);
+					bool isAboveClouds = (cameraPosition.y > cloudAltitude.x) && (cameraPosition.y < cloudAltitude.y || worldDir.y < 0.0);
+					if (isAboveClouds) LandAtmosphericScattering(color, waterDist, vec3(0.0), rayWorldPos, worldDir, materialMask.sky > 0.5);
+					float cloudTransmittance = 1.0;
+					#ifdef ATMO_HORIZON
+						vec3 camera = vec3(0.0, max(cameraPosition.y, ATMO_MIN_ALTITUDE) * 0.001 + atmosphereModel_bottom_radius, 0.0);
+					#else
+						vec3 camera = vec3(0.0, max(cameraPosition.y, 63.0) * 0.001 + atmosphereModel_bottom_radius, 0.0);
+					#endif
+					vec2 noise_1 = BlueNoiseTemporal().xy;
+					float wind = 0.0005 * (windTime * CLOUD_SPEED + 10.0 * FTC_OFFSET);
+					vec3 windDirection = vec3(1.0, -0.05, -0.4) * wind;
+					NubisCumulus(color, worldDir, cloudAltitude, colorShadowlight, colorSkylight, windDirection, camera, noise_1, waterDist, cloudTransmittance);
+					if (!isAboveClouds)
+				#endif
+				LandAtmosphericScattering(color, waterDist, vec3(0.0), rayWorldPos, worldDir, materialMask.sky > 0.5);
+			}
 		#endif
 
 		float globalCloudShadow = 1.0 - wetness * (RAIN_SHADOW * 0.985);
@@ -262,7 +285,6 @@
 		#endif
 	#endif
 
-
 	#ifdef DISABLE_SKY_RENDERING
 		if (depth == 1.0) color = vec3(0.0);
 	#endif
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Final_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Final_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Final_FS.glsl	2026-05-10 15:41:04.343945100 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Final_FS.glsl	2026-05-28 14:31:41.833235100 -0700
@@ -316,6 +316,8 @@
 
 
 
+
+
 #ifdef PT_TRACING_EYE
 
 uniform sampler2D shadowtex0;
@@ -499,7 +501,10 @@
 
 				#ifdef DIMENSION_OVERWORLD
 				#ifdef SUNLIGHT_LEAK_FIX
-					sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+					// sunLighting *= saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+					float originalMask = saturate(hitSkylight * 444.0 + float(isEyeInWater == 1));
+					float eyeSkylight = saturate(float(eyeBrightnessSmooth.y) / 240.0);
+					sunLighting *= mix(1.0, originalMask, saturate(1.0 - eyeSkylight * 10.0));
 				#endif
 				#endif
 
@@ -597,6 +602,44 @@
 
 uniform sampler2D pixelData2D;
 //*/
+// layout (rgba32f) uniform image2D imgFFT;
+uniform sampler2D imgFFT_s;
+
+vec3 getConvBloom(vec2 coord) {
+	vec2 coordScale = vec2(1.0 - float(KERNEL_SIZE) / float(FFT_SIZE));
+
+	// Determine bounding box dimensions based on aspect ratio
+	if (aspectRatio >= 1.0) {
+		coordScale.y /= aspectRatio;
+	} else {
+		coordScale.x *= aspectRatio;
+	}
+	return max(vec3(0.0), texture2D(imgFFT_s, 1.0 - (coord - 2) * pixelSize * coordScale).rgb);
+}
+
+#if defined DIMENSION_OVERWORLD
+	const int ConvBloomStrength = (1 << CONV_BLOOM_STRENGTH);
+#elif defined DIMENSION_END
+	const int ConvBloomStrength = (1 << CONV_BLOOM_STRENGTH_END);
+#else
+	const int ConvBloomStrength = (1 << CONV_BLOOM_STRENGTH_NETHER);
+#endif
+
+#if CAP_LUMINANCE > -1
+  const float maxLuminance = float(1 << CAP_LUMINANCE);
+#endif
+
+// soft knee thresholding
+vec3 softKnee(vec3 color) {
+  float l = Luminance(color.rgb);
+  float x = max(0, l - BLOOM_THRESHOLD + BLOOM_KNEE);
+  color *= mix(x * x / (4 * BLOOM_KNEE), max(0, l - BLOOM_THRESHOLD), float(x > BLOOM_THRESHOLD + BLOOM_KNEE)) / max(l, 1e-6);
+  #if CAP_LUMINANCE > -1
+    color *= min(1.0, maxLuminance / l);
+  #endif
+  return color;
+}
+
 /////////////////////////MAIN//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 /////////////////////////MAIN//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 void main(){
@@ -625,16 +668,31 @@
 		if (hideGUI > 0) color = color * 0.5 + texelFetch(FBTEX_ALBEDO, texelCoord, 0).rgb * 0.5;
 
 	#else
-
-		#ifdef BLOOM
-			color = MergeBloom(color, 1.0 - data1.a);
+		#ifdef DEBUG_CONV_BLOOM
+			color = getConvBloom(texelCoord);
+		#else
+			#ifdef DEBUG_THRESHOLD
+				color = softKnee(color * GetExposureValue());
+			#endif
 		#endif
-
-		#ifdef VIGNETTE
-			color *= Vignette(texCoord, VIGNETTE_FALLOFF, VIGNETTE_ROUNDNESS);
+		#if !defined DEBUG_CONV_BLOOM && !defined DEBUG_THRESHOLD
+			float expValue = GetExposureValue();
+			#if defined EXPOSURE_BASED_BLOOM && defined BLOOM
+			// mix between normal bloom and convolution bloom to simulate aperture adjusting based on brightness
+				color = mix(
+					MergeBloom(color, 1.0 - data1.a) * expValue,
+					color * expValue + 2 * ConvBloomStrength * getConvBloom(texelCoord),
+					min(1.0, EXPOSURE_BLOOM_TRANSITION / expValue)
+				); // 256, 512
+			#elif defined BLOOM && !defined CONV_BLOOM
+				color = MergeBloom(color, 1.0 - data1.a) * expValue;
+			#elif defined CONV_BLOOM
+				color = color * expValue + ConvBloomStrength * getConvBloom(texelCoord);
+			#else
+				color *= expValue;
+			#endif
 		#endif
 
-		color *= GetExposureValue();
 
 		color = TONEMAP_OPERATOR(color);
 
@@ -643,6 +701,10 @@
 		#ifdef ADVANCED_COLOR
 			color = ColorGrading(color);
 		#endif
+
+		#ifdef VIGNETTE
+			color *= Vignette(texCoord, VIGNETTE_FALLOFF, VIGNETTE_ROUNDNESS);
+		#endif
 /*
 		#ifdef HAS_COLORWHEEL
 			#ifdef SUPER_RESOLUTION
@@ -1292,6 +1354,19 @@
 
 	//if (texelCoord.x < 400) color = vec3(1.3);
 
+	#ifdef DEBUG_DEPTH
+	if (all(lessThan(gl_FragCoord.xy, ivec2(100)))) {
+		color = texelFetch(pixelData2D, ivec2(PIXELDATA_CENTER_DEPTH, 0), 0).xxx / 128;
+	} else if (all(lessThan(gl_FragCoord.xy, screenSize / 3))) {
+		ivec2 depthCoord = ivec2(gl_FragCoord.xy * 3 * fsrRenderScale);
+		float depth = texelFetch(depthtex1, depthCoord, 0).x;
+		if (depth == 1.0) {
+			depth = ScreenDepth_From_LODScreenDepth(texelFetch(LOD_DEPTH_TEX_1, depthCoord, 0).x);
+		}
+		color = vec3(2 * near / (far + near - (far - near) * depth));
+	}
+	#endif
+
 	gl_FragData[0] = vec4(color, 0.0);
 
 }
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Gbuffers/VX_Water_FS.glsl 0.8.22-patched/shaders/Lib/Programs/Gbuffers/VX_Water_FS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Gbuffers/VX_Water_FS.glsl	2026-05-10 15:41:05.672022000 -0700
+++ 0.8.22-patched/shaders/Lib/Programs/Gbuffers/VX_Water_FS.glsl	2026-05-10 18:37:22.000000000 -0700
@@ -30,8 +30,9 @@
 //albedo
 	vec4 albedo = parameters.sampledColour * parameters.tinting;
 
+	bool isWater = parameters.customId == 6000.0;
 //lightmap
-	vec2 blockLight = vec2(0.0, saturate(parameters.lightMap.y * 16.0 / 15.0 - 0.5 / 15.0));
+	vec2 blockLight = vec2(0.0, saturate(parameters.lightMap.y * 16.0 / 15.0 - 0.5 / 15.0) * float(isWater));
 
 //TBN
 	vec3 N = vec3(
@@ -47,7 +48,6 @@
 
 //material
 	float materialIDs = MATID_STAINEDGLASS;
-	bool isWater = parameters.customId == 6000.0;
 
 //normal
 	vec3 waterNormal = tbn[2];
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Programs/Setup/KernelFFT_CS.glsl 0.8.22-patched/shaders/Lib/Programs/Setup/KernelFFT_CS.glsl
--- 0.8.22-orig/shaders/Lib/Programs/Setup/KernelFFT_CS.glsl	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/Lib/Programs/Setup/KernelFFT_CS.glsl	2026-05-26 19:07:36.237473800 -0700
@@ -0,0 +1,81 @@
+// by huj31415
+
+// run once on shader load or window resize
+// compute FFT of diffraction kernel and write to image
+
+#include "/Lib/UniformDeclare.glsl"
+#include "/Lib/Utilities.glsl"
+
+#include "/Lib/BasicFunctions/FFT.glsl"
+
+// N rows, N/4 threads per row
+const ivec3 workGroups = ivec3(1, FFT_SIZE, 1);
+layout (local_size_x = N_4_1, local_size_y = 1) in;
+
+#ifdef ROW
+// Row pass ===================================================================
+
+uniform sampler2D ker;
+layout (rgba32f) uniform writeonly image2D tmpFFT;
+
+// pads and scales kernel texture for FFT input
+vec4 sampleCenterKernel(uint x, uint y) {
+  // map thread coordinates to signed distance from the FFT origin
+  int sx = int(x);
+  int sy = int(y);
+  if (sx >= int(FFT_SIZE / 2)) sx -= int(FFT_SIZE);
+  if (sy >= int(FFT_SIZE / 2)) sy -= int(FFT_SIZE);
+
+  // check if thread falls outside the target bloom size
+  int halfK = KERNEL_SIZE / 2;
+  if (abs(sx) >= halfK || abs(sy) >= halfK) {
+    return vec4(0.0);
+  }
+
+  // map the signed offset directly out from the texture center
+  vec2 uv = 0.5 + vec2(sx, sy) / float(KERNEL_SIZE);
+
+  return textureLod(ker, uv, 0.0);
+}
+
+void main() {
+  uint y = gl_GlobalInvocationID.y;
+  uint x = gl_GlobalInvocationID.x;
+
+  vec4 t0 = sampleCenterKernel(x, y);
+  vec4 t1 = sampleCenterKernel(x + N_4_1, y);
+  vec4 t2 = sampleCenterKernel(x + N_4_2, y);
+  vec4 t3 = sampleCenterKernel(x + N_4_3, y);
+
+  rgbaRowFFT(y, x, t0, t1, t2, t3);
+
+  imageStore(tmpFFT, ivec2(y, x),         t0);
+  imageStore(tmpFFT, ivec2(y, x + N_4_1), t1);
+  imageStore(tmpFFT, ivec2(y, x + N_4_2), t2);
+  imageStore(tmpFFT, ivec2(y, x + N_4_3), t3);
+}
+
+#else
+// Column pass ================================================================
+
+layout (rgba32f) uniform image2D tmpFFT;
+layout (rgba32f) uniform writeonly image2D kerFFT;
+
+void main() {
+  uint y = gl_GlobalInvocationID.y;
+  uint x = gl_GlobalInvocationID.x;
+
+  vec4 t0 = imageLoad(tmpFFT, ivec2(x, y));
+  vec4 t1 = imageLoad(tmpFFT, ivec2(x + N_4_1, y));
+  vec4 t2 = imageLoad(tmpFFT, ivec2(x + N_4_2, y));
+  vec4 t3 = imageLoad(tmpFFT, ivec2(x + N_4_3, y));
+
+  rgbaRowFFT(y, x, t0, t1, t2, t3);
+
+  imageStore(kerFFT, ivec2(y, x),         t0 * NORM);
+  imageStore(kerFFT, ivec2(y, x + N_4_1), t1 * NORM);
+  imageStore(kerFFT, ivec2(y, x + N_4_2), t2 * NORM);
+  imageStore(kerFFT, ivec2(y, x + N_4_3), t3 * NORM);
+}
+
+#endif
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/Settings.glsl 0.8.22-patched/shaders/Lib/Settings.glsl
--- 0.8.22-orig/shaders/Lib/Settings.glsl	2026-05-10 15:41:03.462110000 -0700
+++ 0.8.22-patched/shaders/Lib/Settings.glsl	2026-05-28 14:36:13.865134300 -0700
@@ -570,7 +570,7 @@
 
 //Volumetric Fog------------------------
 	#define VFOG
-	#define VFOG_NOISE_TYPE 	0 //[0 1]
+	#define VFOG_NOISE_TYPE 	1 //[0 1]
 
 	#define VFOG_NOISE_OCTAVE 			1  // [1 2 3 4 5 6]
 	#define VFOG_NOISE_HORIZONTAL_SCALE 1.0 // [0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0 1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9 2.0 2.5 3.0 5.0]
@@ -670,6 +670,8 @@
 	#define SMOOTH_EXPOSURE
 	#define EXPOSURE_TIME 	1.0 // [0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0 1.2 1.4 1.6 1.8 2.0]
 
+	#define FLARE_EXPOSURE_COMPENSATION 0.8 // [0.0 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0]
+
 //Motion Blur---------------------------
 	#define MOTION_BLUR
 	#define MOTION_BLUR_DITHER
@@ -691,7 +693,7 @@
 	#define NETHER_END_BLOOM_BOOST 	1.0 // [0.0 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0 1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9 2.0]
 
 	#define LENS_GLARE
-  //#define LENS_FLARE
+  #define LENS_FLARE
 	#define GLARE_BRIGHTNESS 	1.0 // [0.1 0.2 0.3 0.5 0.7 1.0 1.5 2.0 3.0 5.0 7.0 10.0 15.0 20.0]
 	#define FLARE_BRIGHTNESS 	1.0 // [0.1 0.2 0.3 0.5 0.7 1.0 1.5 2.0 3.0 5.0 7.0 10.0 15.0 20.0]
   //#define GLARE_FLARE_SHADOWBASED
@@ -805,6 +807,7 @@
 
 //DEBUG---------------------------------
 	#define WHITE_DEBUG_WORLD 0 // [0 1 2 3 4 5 6 7 8 9 10]
+	//#define DEBUG_DEPTH
   //#define DEBUG_COUNTER
   //#define DISABLE_NIGHTVISION
   //#define DISABLE_BLINDNESS_DARKNESS
@@ -814,4 +817,43 @@
   //#define DEBUG_IRC
   //#define DISABLE_REFLECTION_TEMPORAL_MOTIONWEIGHT
 
-  //#define SHOW_TODO
\ No newline at end of file
+  //#define SHOW_TODO
+
+//PATCHES--------------------------------
+	#define DYNAMIC_WEATHER
+	#define MOON_PHASES
+	#define MOON_PHASE_LIGHTING
+	#define NEW_MOON_BRIGHTNESS 0.005 // [0.001 0.002 0.005 0.01 0.02]
+	#define LEAVES_SSS_MULT 2.5 // [1.0 1.5 2.0 2.5 3.0 4.0 5.0]
+
+	#define KER_TEX_SIZE 1024 // [256 512 1024]
+	#define FFT_SIZE 2048 // [1024 2048]
+// controls how much area the kernel can cover, leaves FFT_SIZE - KERNEL_SIZE pixels for the image
+	#define KERNEL_FRACTION 0.375 // [0.125 0.25 0.375 0.5]
+	const int KERNEL_SIZE = int(FFT_SIZE * KERNEL_FRACTION); // [256 384 512 768]
+
+	#define BLOOM_THRESHOLD 10.0 // [0.1 1.0 2.0 3.0 5.0 7.0 10.0 15.0 20.0]
+	#define BLOOM_KNEE_MULT 0.5 // [0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0]
+	#define CAP_LUMINANCE 5 // [-1 3 4 5 6 7 8 9 10]
+	const int BLOOM_KNEE = int(BLOOM_THRESHOLD * BLOOM_KNEE_MULT); // [0.0 0.2 0.4 0.8 1.0 1.5 2.0 5.0 10.0]
+	#define CONV_BLOOM
+	#define CONV_BLOOM_STRENGTH 5 // [0 1 2 3 4 5 6 7 8 9 10 11 12]
+	#define CONV_BLOOM_STRENGTH_NETHER 4 // [0 1 2 3 4 5 6 7 8 9 10 11 12]
+	#define CONV_BLOOM_STRENGTH_END 2 // [0 1 2 3 4 5 6 7 8 9 10 11 12]
+	#define EXPOSURE_BASED_BLOOM
+	#define EXPOSURE_BLOOM_TRANSITION 10 // [1 5 10 20 50 100 500 1000 5000 10000]
+	//#define DEBUG_THRESHOLD
+	//#define DEBUG_CONV_BLOOM
+
+// require bloom, autoexposure, convolution bloom when enabling exposure based bloom for pass toggling
+#ifdef EXPOSURE_BASED_BLOOM
+	#ifndef BLOOM
+		#define BLOOM
+	#endif
+	#ifdef MANUAL_EXPOSURE
+		#undef MANUAL_EXPOSURE
+	#endif
+	#ifndef CONV_BLOOM
+		#define CONV_BLOOM
+	#endif
+#endif
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/Lib/UniformDeclare.glsl 0.8.22-patched/shaders/Lib/UniformDeclare.glsl
--- 0.8.22-orig/shaders/Lib/UniformDeclare.glsl	2026-05-10 15:41:03.492844600 -0700
+++ 0.8.22-patched/shaders/Lib/UniformDeclare.glsl	2026-05-16 01:24:14.250455600 -0700
@@ -8,10 +8,22 @@
 uniform int heldItemId2;
 uniform int heldBlockLightValue2;
 uniform int worldTime;
+uniform int worldDay;
+
+uniform int moonPhase;
+
+#ifdef MOON_PHASE_LIGHTING
+	float PHASED_NIGHT_BRIGHTNESS = NIGHT_BRIGHTNESS * (pow(abs(4 - moonPhase) / 4.0, 2.2) + float(moonPhase == 4) * NEW_MOON_BRIGHTNESS);
+#else
+	float PHASED_NIGHT_BRIGHTNESS = NIGHT_BRIGHTNESS;
+#endif
+
 uniform int frameCounter;
 uniform float frameTime;
 uniform float frameTimeCounter;
+float windTime = (worldDay % 180 * 1200.0 + worldTime * 0.05);
 uniform float sunAngle;
+uniform float shadowAngle;
 uniform float aspectRatio;
 uniform float viewWidth;
 uniform float viewHeight;
@@ -33,6 +45,31 @@
 uniform float darknessFactor;
 uniform float darknessLightFactor;
 
+#ifdef DYNAMIC_WEATHER
+	vec2 weatherHash(vec2 n) {
+		return fract(sin(n) * 43758.5453);
+	}
+	// daily cumulus cloud coverage centered on CLOUD_CLEAR_COVERY
+	// set max coverage to 1
+	const float coverDeviation = (1 - CLOUD_CLEAR_COVERY);
+	vec2 weather = weatherHash(vec2(worldDay, worldDay + 1)) * 2.0 - 1.0;
+	float weatherBlend = smoothstep(0, 1, worldTime / 24000.0);
+	// blend coverage between days centered around clear coverage setting
+	float DAILY_COVERAGE = clamp(
+		mix(weather.x, weather.y, weatherBlend) * coverDeviation + CLOUD_CLEAR_COVERY, 0.0, 1.0
+	);
+	float FINAL_CLOUD_COVERAGE = mix(DAILY_COVERAGE, CLOUD_RAIN_COVERY, wetness);
+
+	// planar clouds max coverage PC_CLEAR_COVERAGE
+	vec2 PC_weather = weatherHash(vec2(worldDay + 0.5, worldDay + 1.5));
+	float FINAL_PC_COVERAGE = clamp(
+		mix(PC_weather.x, PC_weather.y, weatherBlend) * PC_CLEAR_COVERAGE, 0.1, 1.0
+	);
+#else
+	const float FINAL_CLOUD_COVERAGE = CLOUD_CLEAR_COVERY;
+	const float FINAL_PC_COVERAGE = PC_CLEAR_COVERAGE;
+#endif
+
 uniform vec2 taaJitter;
 uniform vec2 taaJitterToPrevious;
 uniform vec2 previousTaaJitter;
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite77.csh 0.8.22-patched/shaders/composite77.csh
--- 0.8.22-orig/shaders/composite77.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/composite77.csh	2026-05-23 23:31:21.979342700 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Composite/ImageFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite78.csh 0.8.22-patched/shaders/composite78.csh
--- 0.8.22-orig/shaders/composite78.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/composite78.csh	2026-05-23 23:31:23.736721300 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Composite/ImageFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite79.csh 0.8.22-patched/shaders/composite79.csh
--- 0.8.22-orig/shaders/composite79.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/composite79.csh	2026-05-23 23:31:25.430895400 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite80.csh 0.8.22-patched/shaders/composite80.csh
--- 0.8.22-orig/shaders/composite80.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/composite80.csh	2026-05-23 23:31:27.127883100 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite80.fsh 0.8.22-patched/shaders/composite80.fsh
--- 0.8.22-orig/shaders/composite80.fsh	2026-05-10 15:41:02.564122800 -0700
+++ 0.8.22-patched/shaders/composite80.fsh	1969-12-31 16:00:00.000000000 -0800
@@ -1,8 +0,0 @@
-#version 430
-
-
-#define DIMENSION_OVERWORLD
-#define PROGRAM_FINAL_0
-
-
-#include "/Lib/Programs/Final_FS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite80.vsh 0.8.22-patched/shaders/composite80.vsh
--- 0.8.22-orig/shaders/composite80.vsh	2026-05-10 15:41:02.578153600 -0700
+++ 0.8.22-patched/shaders/composite80.vsh	1969-12-31 16:00:00.000000000 -0800
@@ -1,7 +0,0 @@
-#version 430
-
-
-#define DIMENSION_OVERWORLD
-
-
-#include "/Lib/Programs/Composite/PixelData_VS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite81.fsh 0.8.22-patched/shaders/composite81.fsh
--- 0.8.22-orig/shaders/composite81.fsh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/composite81.fsh	2026-05-10 18:37:20.000000000 -0700
@@ -0,0 +1,8 @@
+#version 430
+
+
+#define DIMENSION_OVERWORLD
+#define PROGRAM_FINAL_0
+
+
+#include "/Lib/Programs/Final_FS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/composite81.vsh 0.8.22-patched/shaders/composite81.vsh
--- 0.8.22-orig/shaders/composite81.vsh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/composite81.vsh	2026-05-10 18:37:20.000000000 -0700
@@ -0,0 +1,7 @@
+#version 430
+
+
+#define DIMENSION_OVERWORLD
+
+
+#include "/Lib/Programs/Composite/PixelData_VS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/lang/en_us.lang 0.8.22-patched/shaders/lang/en_us.lang
--- 0.8.22-orig/shaders/lang/en_us.lang	2026-05-10 15:41:03.397818800 -0700
+++ 0.8.22-patched/shaders/lang/en_us.lang	2026-05-26 19:27:38.002142400 -0700
@@ -779,3 +779,41 @@
 
 
 screen.DEBUG=§7Debug
+
+option.FLARE_EXPOSURE_COMPENSATION=Flare Exposure Compensation
+option.LENS_FLARE_MODE=Lens Flare Style
+value.LENS_FLARE_MODE.0=Camera (Anamorphic)
+value.LENS_FLARE_MODE.1=Human Eye (Entoptic)
+option.FLARE_GHOSTING=Ghosting
+option.GLARE_FLARE_SHADOWBASED=Based on Shadow
+
+screen.PATCHES=huj31415's patches
+option.DYNAMIC_WEATHER=Dynamic cloud cover
+option.DYNAMIC_WEATHER.comment=Vary cloud cover between days
+option.MOON_PHASE_LIGHTING=Moon phase-based night brightness
+option.NEW_MOON_BRIGHTNESS=New moon night brightness
+option.LEAVES_SSS_MULT=Leaves subsurface scattering multiplier
+
+screen.CONVOLUTION_BLOOM=Convolution bloom
+option.FFT_SIZE=FFT size
+option.FFT_SIZE.comment=Higher can resolve smaller lights but is a bit slower
+option.KERNEL_FRACTION=Kernel size
+option.KERNEL_FRACTION.comment=Fraction of the FFT size covered by the bloom kernel. Larger values result in blooms that cover more of the screen but lower bloom image resolution (only signficiant for FFT size 1024)
+option.KER_TEX_SIZE=Kernel texture size §c(Only use for custom kernel textures)
+option.KER_TEX_SIZE.comment=Controls the kernel source texture size. Does not affect the size of the bloom since the texture is sampled to the kernel size above, but resolutions less than 512 may affect the bloom's appearance. 
+option.CONV_BLOOM=Enabled
+option.CONV_BLOOM_STRENGTH=Strength (exp2)
+option.CONV_BLOOM_STRENGTH_NETHER=Strength in Nether (exp2)
+option.CONV_BLOOM_STRENGTH_END=Strength in End (exp2)
+option.CAP_LUMINANCE=Cap luminance (exp2)
+option.CAP_LUMINANCE.comment=Set to -1 to disable. Reduces dynamic range input into FFT to allow artificial light sources to bloom without the Sun becoming blown out. Enable for noticeable bloom in the dark, disable and turn down strength for the most physical result
+option.BLOOM_THRESHOLD=Bloom threshold
+option.BLOOM_THRESHOLD.comment=Minimum luminance to affect bloom
+option.BLOOM_KNEE_MULT=Threshold soft knee
+option.BLOOM_KNEE_MULT.comment=Width of threshold soft knee relative to threshold
+option.EXPOSURE_BASED_BLOOM=Exposure based bloom
+option.EXPOSURE_BASED_BLOOM.comment=Transition between convolution bloom diffraction spikes at lower exposures and the original Gaussian bloom at higher exposures to simulate changing aperture size. Will force bloom, convolution bloom, and autoexposure
+option.EXPOSURE_BLOOM_TRANSITION=Bloom transition exposure
+option.EXPOSURE_BLOOM_TRANSITION.comment=Exposure where the bloom starts transitioning between convolution and Gaussian. Higher values begin transition in darker environments
+option.DEBUG_THRESHOLD=DEBUG Thresholded scene
+option.DEBUG_CONV_BLOOM=DEBUG Convolved image
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/setup.csh 0.8.22-patched/shaders/setup.csh
--- 0.8.22-orig/shaders/setup.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/setup.csh	2026-05-24 19:46:39.967472300 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Setup/KernelFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/setup1.csh 0.8.22-patched/shaders/setup1.csh
--- 0.8.22-orig/shaders/setup1.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/setup1.csh	2026-05-24 19:46:16.950890100 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Setup/KernelFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/shaders.properties 0.8.22-patched/shaders/shaders.properties
--- 0.8.22-orig/shaders/shaders.properties	2026-05-10 15:41:03.265972000 -0700
+++ 0.8.22-patched/shaders/shaders.properties	2026-05-28 14:37:04.151967200 -0700
@@ -60,6 +60,27 @@
 program.world1/composite90.enabled=WAVEFORM_SCOPE
 program.world-1/composite90.enabled=WAVEFORM_SCOPE
 
+program.setup.enabled=CONV_BLOOM
+program.setup1.enabled=CONV_BLOOM
+program.composite77.enabled=CONV_BLOOM
+program.composite78.enabled=CONV_BLOOM
+program.composite79.enabled=CONV_BLOOM
+program.composite80.enabled=CONV_BLOOM
+
+program.world1/setup.enabled=CONV_BLOOM
+program.world1/setup1.enabled=CONV_BLOOM
+program.world1/composite77.enabled=CONV_BLOOM
+program.world1/composite78.enabled=CONV_BLOOM
+program.world1/composite79.enabled=CONV_BLOOM
+program.world1/composite80.enabled=CONV_BLOOM
+
+program.world-1/setup.enabled=CONV_BLOOM
+program.world-1/setup1.enabled=CONV_BLOOM
+program.world-1/composite77.enabled=CONV_BLOOM
+program.world-1/composite78.enabled=CONV_BLOOM
+program.world-1/composite79.enabled=CONV_BLOOM
+program.world-1/composite80.enabled=CONV_BLOOM
+
 
 
 flip.composite10.colortex0=true
@@ -78,15 +99,15 @@
 flip.composite40.colortex9=true
 flip.composite45.colortex9=true
 flip.composite67.colortex9=true
-flip.composite80.colortex9=true
+flip.composite81.colortex9=true
 flip.world1/composite.colortex9=true
 flip.world1/composite45.colortex9=true
 flip.world1/composite67.colortex9=true
-flip.world1/composite80.colortex9=true
+flip.world1/composite81.colortex9=true
 flip.world-1/composite.colortex9=true
 flip.world-1/composite45.colortex9=true
 flip.world-1/composite67.colortex9=true
-flip.world-1/composite80.colortex9=true
+flip.world-1/composite81.colortex9=true
 
 flip.composite25.colortex10=true
 flip.composite67.colortex10=true
@@ -435,6 +456,15 @@
 #image.img_rippleXY2D = rippleXY2D RG RG32F FLOAT false false 2048 2048
 #image.img_ripple2D = ripple2D RGBA RGBA8 UNSIGNED_BYTE false false 128 7680
 
+customTexture.ker = texture/kernelTex.bin TEXTURE_2D RGBA32F KER_TEX_SIZE KER_TEX_SIZE RGBA FLOAT
+
+image.tmpFFT = tmpFFT_s RGBA RGBA32F FLOAT true false FFT_SIZE FFT_SIZE
+
+image.kerFFT = kerFFT_s RGBA RGBA32F FLOAT false false FFT_SIZE FFT_SIZE
+
+image.imgFFT = imgFFT_s RGBA RGBA32F FLOAT true false FFT_SIZE FFT_SIZE
+
+
 
 texture.noise=texture/noise.png
 
@@ -809,7 +839,8 @@
 screen.SCREENSHOT.columns=1
 	screen.ANIMATION.columns=2
 screen.DEBUG.columns=2
-
+screen.PATCHES.columns=1
+	screen.CONVOLUTION_BLOOM.columns=1
 
 
 
@@ -834,7 +865,7 @@
 		[SCREENSHOT]        <empty> \
 		<empty>             <empty> \
 		<empty>             <empty> \
-		[DEBUG]
+		[DEBUG]             [PATCHES]
 
 screen.PT_VOXEL=PT_VOXEL_RESOLUTION <empty> PT_MIDBLOCK_TEMPFIX PT_FULLBLOCK_VERIFICATION <empty> PT_TAG_DETECTION PT_FULLBLOCK_DETECTION PT_FULLBLOCK_DETECTION_ALPHA_TRACING <empty> VANILLA_EMISSIVE
 
@@ -1115,7 +1146,7 @@
 							AE_OFFSET                 <empty> \
 							<empty>                   MANUAL_EXPOSURE \
 							LUMINANCE_WEIGHT_MODE     EV_VALUE \
-							LUMINANCE_WEIGHT_STRENGTH
+							LUMINANCE_WEIGHT_STRENGTH FLARE_EXPOSURE_COMPENSATION
 
 	screen.MOTION_BLUR_SETTING= MOTION_BLUR MOTION_BLUR_DITHER MOTION_BLUR_QUALITY <empty> MOTION_BLUR_SUTTER_MODE MOTION_BLUR_SUTTER_ANGLE MOTION_BLUR_SUTTER_SPEED
 
@@ -1174,7 +1205,7 @@
 
 screen.DEBUG =  DEBUG_COUNTER              PT_TRACING_EYE \
 				WHITE_DEBUG_WORLD          DEBUG_IRC \
-				<empty>                    PT_TEMP_COORD_OFFSET \
+				DEBUG_DEPTH                PT_TEMP_COORD_OFFSET \
 				FTC_OFFSET                 <empty> \
 				CLOUD_SPEED                DEBUG_RTW \
 				<empty>                    <empty> \
@@ -1183,6 +1214,21 @@
 				DISABLE_REFLECTION_TEMPORAL_MOTIONWEIGHT
 
 
+screen.PATCHES =  DYNAMIC_WEATHER     MOON_PHASES \
+								MOON_PHASE_LIGHTING   NEW_MOON_BRIGHTNESS \
+								LEAVES_SSS_MULT       <empty> \
+								[CONVOLUTION_BLOOM]
+	screen.CONVOLUTION_BLOOM = CONV_BLOOM     FFT_SIZE \
+								<empty> \
+	              KER_TEX_SIZE          KERNEL_FRACTION \
+								<empty> \
+								CONV_BLOOM_STRENGTH CONV_BLOOM_STRENGTH_NETHER CONV_BLOOM_STRENGTH_END \
+								BLOOM_THRESHOLD       BLOOM_KNEE_MULT  CAP_LUMINANCE \
+								<empty> \
+								EXPOSURE_BASED_BLOOM  EXPOSURE_BLOOM_TRANSITION \
+								BLOOM \
+								<empty> \
+								DEBUG_THRESHOLD       DEBUG_CONV_BLOOM
 
 
 
@@ -1228,4 +1274,6 @@
 		DH_TEXTURE_NOISE_STEPS DH_TEXTURE_NOISE_STRENGTH \
 		TACZ_ADAPTIVE_EMISSIVE_BRIGHTNESS TACZ_MUZZLE_FLASH_BRIGHTNESS \
 		RENDERING_MODE_SPP RENDERING_MODE_COOLDOWN \
-		WHITE_DEBUG_WORLD FTC_OFFSET
+		WHITE_DEBUG_WORLD FTC_OFFSET \
+		NEW_MOON_BRIGHTNESS LEAVES_SSS_MULT \
+		KER_TEX_SIZE FFT_SIZE KERNEL_FRACTION CONV_BLOOM_STRENGTH CONV_BLOOM_STRENGTH_NETHER CONV_BLOOM_STRENGTH_END BLOOM_THRESHOLD BLOOM_KNEE_MULT CAP_LUMINANCE EXPOSURE_BLOOM_TRANSITION
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/texture/kernelTex.bin.mcmeta 0.8.22-patched/shaders/texture/kernelTex.bin.mcmeta
--- 0.8.22-orig/shaders/texture/kernelTex.bin.mcmeta	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/texture/kernelTex.bin.mcmeta	2026-05-23 22:56:26.568689900 -0700
@@ -0,0 +1,6 @@
+{
+  "texture": {
+    "blur": true,
+    "clamp": false
+  }
+}
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite77.csh 0.8.22-patched/shaders/world-1/composite77.csh
--- 0.8.22-orig/shaders/world-1/composite77.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/composite77.csh	2026-05-23 23:30:51.348297000 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Composite/ImageFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite78.csh 0.8.22-patched/shaders/world-1/composite78.csh
--- 0.8.22-orig/shaders/world-1/composite78.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/composite78.csh	2026-05-23 23:30:53.565218100 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Composite/ImageFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite79.csh 0.8.22-patched/shaders/world-1/composite79.csh
--- 0.8.22-orig/shaders/world-1/composite79.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/composite79.csh	2026-05-23 23:30:56.274458500 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite80.csh 0.8.22-patched/shaders/world-1/composite80.csh
--- 0.8.22-orig/shaders/world-1/composite80.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/composite80.csh	2026-05-23 23:30:58.265386600 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite80.fsh 0.8.22-patched/shaders/world-1/composite80.fsh
--- 0.8.22-orig/shaders/world-1/composite80.fsh	2026-05-10 15:41:07.516792500 -0700
+++ 0.8.22-patched/shaders/world-1/composite80.fsh	1969-12-31 16:00:00.000000000 -0800
@@ -1,8 +0,0 @@
-#version 430
-
-
-#define DIMENSION_NETHER
-#define PROGRAM_FINAL_0
-
-
-#include "/Lib/Programs/Final_FS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite80.vsh 0.8.22-patched/shaders/world-1/composite80.vsh
--- 0.8.22-orig/shaders/world-1/composite80.vsh	2026-05-10 15:41:07.531313800 -0700
+++ 0.8.22-patched/shaders/world-1/composite80.vsh	1969-12-31 16:00:00.000000000 -0800
@@ -1,7 +0,0 @@
-#version 430
-
-
-#define DIMENSION_NETHER
-
-
-#include "/Lib/Programs/Composite/PixelData_VS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite81.fsh 0.8.22-patched/shaders/world-1/composite81.fsh
--- 0.8.22-orig/shaders/world-1/composite81.fsh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/composite81.fsh	2026-05-10 18:37:20.000000000 -0700
@@ -0,0 +1,8 @@
+#version 430
+
+
+#define DIMENSION_NETHER
+#define PROGRAM_FINAL_0
+
+
+#include "/Lib/Programs/Final_FS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/composite81.vsh 0.8.22-patched/shaders/world-1/composite81.vsh
--- 0.8.22-orig/shaders/world-1/composite81.vsh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/composite81.vsh	2026-05-10 18:37:20.000000000 -0700
@@ -0,0 +1,7 @@
+#version 430
+
+
+#define DIMENSION_NETHER
+
+
+#include "/Lib/Programs/Composite/PixelData_VS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/setup.csh 0.8.22-patched/shaders/world-1/setup.csh
--- 0.8.22-orig/shaders/world-1/setup.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/setup.csh	2026-05-24 19:46:51.653385800 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Setup/KernelFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world-1/setup1.csh 0.8.22-patched/shaders/world-1/setup1.csh
--- 0.8.22-orig/shaders/world-1/setup1.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world-1/setup1.csh	2026-05-24 19:46:21.734050300 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Setup/KernelFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite77.csh 0.8.22-patched/shaders/world1/composite77.csh
--- 0.8.22-orig/shaders/world1/composite77.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/composite77.csh	2026-05-23 23:31:04.364655000 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Composite/ImageFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite78.csh 0.8.22-patched/shaders/world1/composite78.csh
--- 0.8.22-orig/shaders/world1/composite78.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/composite78.csh	2026-05-23 23:31:06.010123200 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Composite/ImageFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite79.csh 0.8.22-patched/shaders/world1/composite79.csh
--- 0.8.22-orig/shaders/world1/composite79.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/composite79.csh	2026-05-23 23:31:10.295136700 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite80.csh 0.8.22-patched/shaders/world1/composite80.csh
--- 0.8.22-orig/shaders/world1/composite80.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/composite80.csh	2026-05-23 23:31:15.012989100 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Composite/ImageConvolveIFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite80.fsh 0.8.22-patched/shaders/world1/composite80.fsh
--- 0.8.22-orig/shaders/world1/composite80.fsh	2026-05-10 15:41:09.574095800 -0700
+++ 0.8.22-patched/shaders/world1/composite80.fsh	1969-12-31 16:00:00.000000000 -0800
@@ -1,8 +0,0 @@
-#version 430
-
-
-#define DIMENSION_END
-#define PROGRAM_FINAL_0
-
-
-#include "/Lib/Programs/Final_FS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite80.vsh 0.8.22-patched/shaders/world1/composite80.vsh
--- 0.8.22-orig/shaders/world1/composite80.vsh	2026-05-10 15:41:09.587795100 -0700
+++ 0.8.22-patched/shaders/world1/composite80.vsh	1969-12-31 16:00:00.000000000 -0800
@@ -1,7 +0,0 @@
-#version 430
-
-
-#define DIMENSION_END
-
-
-#include "/Lib/Programs/Composite/PixelData_VS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite81.fsh 0.8.22-patched/shaders/world1/composite81.fsh
--- 0.8.22-orig/shaders/world1/composite81.fsh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/composite81.fsh	2026-05-10 18:37:20.000000000 -0700
@@ -0,0 +1,8 @@
+#version 430
+
+
+#define DIMENSION_END
+#define PROGRAM_FINAL_0
+
+
+#include "/Lib/Programs/Final_FS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/composite81.vsh 0.8.22-patched/shaders/world1/composite81.vsh
--- 0.8.22-orig/shaders/world1/composite81.vsh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/composite81.vsh	2026-05-10 18:37:20.000000000 -0700
@@ -0,0 +1,7 @@
+#version 430
+
+
+#define DIMENSION_END
+
+
+#include "/Lib/Programs/Composite/PixelData_VS.glsl"
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/setup.csh 0.8.22-patched/shaders/world1/setup.csh
--- 0.8.22-orig/shaders/world1/setup.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/setup.csh	2026-05-24 19:46:28.272019900 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define ROW
+
+#include "/Lib/Programs/Setup/KernelFFT_CS.glsl"
\ No newline at end of file
diff -Nrua --strip-trailing-cr '--exclude=.git' 0.8.22-orig/shaders/world1/setup1.csh 0.8.22-patched/shaders/world1/setup1.csh
--- 0.8.22-orig/shaders/world1/setup1.csh	1969-12-31 16:00:00.000000000 -0800
+++ 0.8.22-patched/shaders/world1/setup1.csh	2026-05-24 19:46:31.572278500 -0700
@@ -0,0 +1,5 @@
+#version 430
+
+#define COL
+
+#include "/Lib/Programs/Setup/KernelFFT_CS.glsl"
\ No newline at end of file
